python/ray/air/util/tensor_extensions/arrow.py

```diff
@@ -188,6 +188,25 @@ def _convert_to_pyarrow_native_array(
         #       by int32 offsets used by Arrow internally)
         dtype = _infer_pyarrow_type(column_values)
 
+        if pa.types.is_timestamp(dtype):
+            assert np.issubdtype(column_values.dtype, np.datetime64)
+            numpy_precision, _ = np.datetime_data(column_values.dtype)
+            arrow_precision = dtype.unit
+            if arrow_precision != numpy_precision:
+                # Arrow supports fewer timestamp resolutions than NumPy. So, if Arrow
+                # doesn't support the resolution, we need to cast the NumPy array to a
+                # different type. This can be a lossy conversion.
+                column_values = column_values.astype(f"datetime64[{arrow_precision}]")
+
+                if log_once(f"column_{column_name}_timestamp_warning"):
+                    logger.warning(
+                        f"Converting a {numpy_precision!r} precision datetime NumPy "
+                        f"array to '{arrow_precision}' precision Arrow timestamp. This "
+                        "conversion occurs because Arrow supports fewer precisions "
+                        "than Arrow and might result in a loss of precision or "
+                        "unrepresentable values."
+                    )
```

Comment:
@bveeramani can you help me understand how's this related to memory-leak?

Reply:
Discussed offline